torchsde/_core/methods_strat/diagonal/heun.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from torchsde._core import base_solver
+
+
+class HeunDiagonal(base_solver.GenericSDESolver):
+
+    def step(self, t0, y0, dt):
+        assert dt > 0, 'Underflow in dt {}'.format(dt)
+
+        I_k = [(bm_next - bm_cur).to(y0[0]) for bm_next, bm_cur in zip(self.bm(t0 + dt), self.bm(t0))]
+
+        f_eval = self.sde.f(t0, y0)
+        g_eval = self.sde.g(t0, y0)
+        g_prod_eval = self.sde.g_prod(g_eval, I_k)
+        y1 = [
+            y0_i + f_eval_i * dt + g_prod_eval_i
+            for y0_i, f_eval_i, g_prod_eval_i in zip(y0, f_eval, g_prod_eval)
+        ]
+        
+        t1 = t0 + dt
+        I_k = [(bm_next - bm_cur).to(y1[0]) for bm_next, bm_cur in zip(self.bm(t1 + dt), self.bm(t1))]
+
+        f_eval_next_step_sum = [
+            a + b for a, b in zip(f_eval, self.sde.f(t1, y1))
+        ]
+        g_eval_next_step_sum = [
+            (a + b) / 2.0 for a,b in zip(g_eval, self.sde.g(t1, y1))
+        ]
+        g_prod_eval_next_step = self.sde.g_prod(g_eval_next_step_sum, I_k)
+        y2 = [
+            y0_i + 0.5 * f_eval_i * dt + g_prod_eval_i
+            for y0_i, f_eval_i, g_prod_eval_i in zip(y0, f_eval_next_step_sum, g_prod_eval_next_step)
+        ]
+
+        return t1, y2
+
+    @property
+    def strong_order(self):
+        return 0.5
```

Comment:
This should be 1.0.

I notice that you get 0.5 in your tests; this is probably because of the earlier bug with the Brownian motion. (Or perhaps because you're comparing Ito against Stratonovich?)

It doesn't help btw that there are actually two different methods called "Heun". One with strong order 0.5, and one with strong order 1.0. The one that it looks like you're aiming to implement has strong order 1.0.